jlm/llvm/opt/alias-analyses/PrecisionEvaluator.hpp

```diff
@@ -0,0 +1,234 @@
+ * See COPYING for terms of redistribution.
+ */
+
+#ifndef JLM_LLVM_OPT_ALIAS_ANALYSES_PRECISIONEVALUATOR_HPP
+#define JLM_LLVM_OPT_ALIAS_ANALYSES_PRECISIONEVALUATOR_HPP
+
+#include <jlm/llvm/opt/alias-analyses/PointsToGraph.hpp>
+#include <jlm/rvsdg/RvsdgModule.hpp>
+#include <jlm/util/Statistics.hpp>
+
+#include <unordered_map>
+
+namespace jlm::llvm::aa
+{
+
+/**
+ * Interface for making alias analysis queries about pairs of pointers
+ */
+class PairwiseAliasAnalysis
+{
+public:
+  PairwiseAliasAnalysis();
+  virtual ~PairwiseAliasAnalysis();
+
+  /**
+   * If this alias analysis is unable to determine NoAlias, it queries the optional backup analysis.
+   * @param backup the PairwiseAliasAnalysis instance to use as backup
+   */
+  void
+  SetBackup(PairwiseAliasAnalysis * backup) noexcept
+  {
+    Backup_ = backup;
+  }
+
+  PairwiseAliasAnalysis *
+  GetBackup() const noexcept
+  {
+    return Backup_;
+  }
+
+  /**
+   * Checks if the pointers represented by p1 and p2 may alias.
+   * If this analysis is unable to prove No Alias, and a backup is provided, the backup is asked.
+   * @param p1 the first pointer value
+   * @param p2 the second pointer value
+   * @return false if the analysis is able to prove that they do not alias, true otherwise
+   */
+  bool
+  MayAlias(const rvsdg::output & p1, const rvsdg::output & p2);
+
+  /**
+   * @return a string description of the pairwise alias analysis, including any backup instances
+   */
+  [[nodiscard]] std::string
+  ToString() const;
+
+protected:
+  virtual bool
+  MayAliasImpl(const rvsdg::output & p1, const rvsdg::output & p2) = 0;
+
+  virtual std::string
+  ToStringImpl() const = 0;
+
+private:
+  // Another instance of PairwiseAliasAnalysis, queried if the current analysis says "May Alias"
+  PairwiseAliasAnalysis * Backup_ = nullptr;
+};
+
+/**
+ * Interface for making alias pairwise analysis queries to a PointsToGraph
+ */
+class PointsToGraphAliasAnalysis final : public PairwiseAliasAnalysis
+{
+public:
+  explicit PointsToGraphAliasAnalysis(PointsToGraph & pointsToGraph);
+  ~PointsToGraphAliasAnalysis() override;
+
+protected:
+  bool
+  MayAliasImpl(const rvsdg::output & p1, const rvsdg::output & p2) override;
+
+  std::string
+  ToStringImpl() const override;
+
+private:
+  // The PointsToGraph used to answer alias queries
+  PointsToGraph & PointsToGraph_;
+};
+
+/**
+ * Option configuring how pointers are collected and evaluated for may-alias precision
+ */
+enum class PrecisionEvaluationMode
+{
+  // In this mode, a pointer use is a store / load, and a clobber is a store
+  ClobberingStores,
+
+  // In this mode, all pointer outputs are both used, and considered to clobber.
+  // This is equivalent to checking all pairs of pointers in each function.
+  // Temporary pointers, such as the result of a pointer offset, are counted as separate pointers.
+  // This is consistent with Lattner's 2007 DSA paper.
+  AllPointerPairs
+};
```

Comment:
Might be subjective, but I would move this into the PrecisionEvaluator class and just call it Mode.